kernel/aster-nix/src/net/iface/common.rs

```diff
@@ -147,56 +149,93 @@ impl IfaceCommon {
+                    break;
+                };
+                if instant > timestamp {
+                    break;
                 }
-            });
-        }
+            }
+            (has_events, poll_at)
+        };
+
+        // drop sockets here to avoid deadlock
+        drop(sockets);
+        drop(interface);
 
-        let sockets = self.sockets.lock_irq_disabled();
-        if let Some(instant) = interface.poll_at(timestamp, &sockets) {
-            let old_instant = self.next_poll_at_ms.load(Ordering::Acquire);
+        if let Some(instant) = poll_at {
+            let old_instant = self.next_poll_at_ms.load(Ordering::Relaxed);
             let new_instant = instant.total_millis() as u64;
             self.next_poll_at_ms.store(new_instant, Ordering::Relaxed);
 
-            if new_instant < old_instant {
+            if old_instant == 0 || new_instant < old_instant {
                 self.polling_wait_queue.wake_all();
             }
         } else {
             self.next_poll_at_ms.store(0, Ordering::Relaxed);
         }
+
+        if has_events {
+            self.bound_sockets.read().iter().for_each(|bound_socket| {
+                bound_socket.on_iface_events();
+            });
+
+            let closed_sockets = self
+                .closing_sockets
+                .lock()
+                .extract_if(|closing_socket| closing_socket.is_closed())
+                .collect::<Vec<_>>();
+            drop(closed_sockets);
+        }
     }
 
     pub(super) fn next_poll_at_ms(&self) -> Option<u64> {
-        let millis = self.next_poll_at_ms.load(Ordering::SeqCst);
+        let millis = self.next_poll_at_ms.load(Ordering::Relaxed);
         if millis == 0 {
             None
         } else {
             Some(millis)
         }
     }
 
-    fn insert_bound_socket(&self, socket: &Arc<AnyBoundSocket>) -> Result<()> {
-        let weak_ref = KeyableWeak::from(Arc::downgrade(socket));
-        let mut bound_sockets = self.bound_sockets.write();
-        if bound_sockets.contains(&weak_ref) {
-            return_errno_with_message!(Errno::EINVAL, "the socket is already bound");
-        }
-        bound_sockets.insert(weak_ref);
-        Ok(())
+    fn insert_bound_socket(&self, socket: &Arc<AnyBoundSocketInner>) {
+        let keyable_socket = KeyableArc::from(socket.clone());
+
+        let inserted = self.bound_sockets.write().insert(keyable_socket);
+        assert!(inserted);
     }
 
-    pub(super) fn remove_bound_socket(&self, socket: Weak<AnyBoundSocket>) {
-        let weak_ref = KeyableWeak::from(socket);
-        self.bound_sockets.write().remove(&weak_ref);
+    pub(super) fn remove_bound_socket_now(&self, socket: &Arc<AnyBoundSocketInner>) {
+        let keyable_socket = KeyableArc::from(socket.clone());
+
+        let removed = self.bound_sockets.write().remove(&keyable_socket);
+        assert!(removed);
+    }
+
+    pub(super) fn remove_bound_socket_when_closed(&self, socket: &Arc<AnyBoundSocketInner>) {
```

Comment:
Is it possible to merge `remove_bound_socket_now` and `remove_bound_socket_when_closed` as one? It looks like the content of `remove_bound_socket_now` is just part of `remove_bound_socket_when_closed`?

A possible solution maybe
```rust
pub(super) fn remove_bound_socket(
    &self, 
    socket: &Arc<AnyBoundSocketInner>, 
    check_closed: impl Fn(&AnyBoundSocketInner) -> bool) {
```

Reply:
> Meanwhile, remove_bound_socket_now() does not need to take the closing_sockets lock. That's the real difference between remove_bound_socket_now() and remove_bound_socket_when_closed()

I neglect that `remove_bound_socket_when_closed` holds the lock before checking the socket status. I can't think up of a better method. And the current from is acceptable for me.